flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -611,27 +611,21 @@ protected void cleanUpInvoke() throws Exception {
 		Thread.interrupted();
 
 		// stop all timers and threads
-		tryShutdownTimerService();
+		Exception suppressedException = suppressThrowable(this::tryShutdownTimerService, null);
 
 		// stop all asynchronous checkpoint threads
-		try {
-			cancelables.close();
-			shutdownAsyncThreads();
-		} catch (Throwable t) {
-			// catch and log the exception to not replace the original exception
-			LOG.error("Could not shut down async checkpoint threads", t);
-		}
+		suppressedException = suppressThrowable(cancelables::close, suppressedException);
+		suppressedException = suppressThrowable(this::shutdownAsyncThreads, suppressedException);
 
 		// we must! perform this cleanup
-		try {
-			cleanup();
-		} catch (Throwable t) {
-			// catch and log the exception to not replace the original exception
-			LOG.error("Error during cleanup of stream task", t);
-		}
+		suppressedException = suppressThrowable(this::cleanup, suppressedException);
 
 		// if the operators were not disposed before, do a hard dispose
-		disposeAllOperators(true);
+		try {
+			disposeAllOperators();
+		} catch (Exception t) {
+			suppressedException = ExceptionUtils.firstOrSuppressed(t, suppressedException);
+		}
 
 		// release the output resources. this method should never fail.
 		if (operatorChain != null) {
```

Comment:
I think the above exceptions can be lost if a new exception is thrown by this block.

Again, why not to use `suppressThrowable` here too? 
To preserve the current behavior (exit immediately if this block fails) we can `throw suppressedException` right after it.
However, I don't think that the current behavior is intentional.

The same about `mailboxProcessor.close` below.

Reply:
What part of the code are you mentioning? `operatorChain`?

I think we agreed in the previous conversation that we want to keep this part and `mailboxProcessor.close()` as it is without changing the behavior, aren't we?

Remember the put all the runnable into a list? I think we have discussed this problem there?

Reply:
Yes, I remember we agreed to preserve the behavior. But this behavior can be implemented differently. And the current implementation loses the above exception (which was logged previously).

Reply:
> Yes, I remember we agreed to preserve the behavior. But this behavior can be implemented differently. And the current implementation loses the above exception (which was logged previously).

Sorry I do not get the comment…. What I am expected to do here? I remember we agreed during our first conversation that we want to keep it as it is, which I think it is as it is now.
I do not get from the comment what I am expected to do….

Reply:
In the current `master`, if an exception is thrown in both `shutdownAsyncThreads` **and** `operatorChain.releaseOutputs` then both exceptions are logged/thrown.

In this PR, only the latter will be thrown.

The simplest fix is to add `if (suppressedException != null) throw suppressedException;` before `if (operatorChain != null)` and `mailboxProcessor.close`.

More elegant way is to add `fail-immediately` flag to `runAndSuppressThrowable` with `true` for lines 629 and 640.

Reply:
That's true. 

We will lose all the `suppressedException` if thrown in the middle of the block you mentioned.

Reply:
Then the question is in this case, who will be the suppressed exception?

The previous behavior actually throws the exception from the block and log the suppressed ones.

Let's quicky sync up on this?